realsense2_camera/src/base_realsense_node.cpp

```diff
@@ -1103,120 +1100,48 @@ static void output_unit_vector(unsigned int height, unsigned int width, float un
-		{
-			for (int x = 0; x < depth_intrinsics.width; ++x)
-			{
-				scaled_depth = static_cast<float>(*image_depth16) * _depth_scale_meters;
-
-				++image_depth16;
-
-				if (fabs(scaled_depth) < 0.0001) {
-					num_bad_pixel++;
-					continue;
-				}
-
-				if (valid_pixels[y][x]) {
-					continue;
-				}
-
-				unit_vectors[y][x * 3] = (x - depth_intrinsics.ppx) / (depth_intrinsics.fx * scaled_depth);
-				unit_vectors[y][x * 3 + 1] = (y - depth_intrinsics.ppy) / (depth_intrinsics.fy * scaled_depth);
-				unit_vectors[y][x * 3 + 2] = 1;
-
-				valid_pixels[y][x] = true;
-				num_valid++;
-			}
-		}
-
-		/* check the % of pixels with valid depth value in this frame */
-		if ((detailed_logging) && (max_log > 0))
-			ROS_INFO("Brain: %% of pixels in this frame with valid depth values is:%f %%.\n",
-						1 - num_bad_pixel / (depth_intrinsics.height * depth_intrinsics.width));
-
-		/* check the % of valid vectors in unit vectors array */
-		float rate = num_valid / (depth_intrinsics.height * depth_intrinsics.width);
-		if (rate >= mature_threshold) {
-			unit_vector_mature = true;
-			ROS_INFO("Brain: %% of pixels with valid depth values is:%f %% which is above the threshold %f %% .\n",
-						rate, mature_threshold);
-
-			ROS_INFO("Brain: prepare the data to be formatted to image message.\n");
-
-			for (unsigned int i = 0; i < height; i++)
-				for (unsigned int j = 0; j < width * 3; j++)
-					current_unit_vectors[i][j] = unit_vectors[i][j];
-
-			if (enable_unit_test)
-				output_unit_vector(height, width, unit_vectors);
-		} else {
-			if ((detailed_logging) && (max_log > 0))
-				ROS_INFO("Brain: %% of pixels with valid depth values is:%f %% which is below the threshold %f %% .\n",
-							rate, mature_threshold);
-		}
-	}
-
-	if (unit_vector_mature || !publish_high_quality) {
-		float any_size_unit_vectors[height][width * 3]= {0};
-		void *data;
-
-		if ((height == FRAME_PRIME_HEIGHT) && (width == FRAME_PRIME_WIDTH)) {
-			data = current_unit_vectors;
-		} else {
-			for (unsigned int i = 0; i < height; i++)
-				for (unsigned int j = 0; j < width * 3; j++)
-					any_size_unit_vectors[i][j] = unit_vectors[i][j];
-
-			data = any_size_unit_vectors;
-		}
+    try
+    {
+        if (!is_frame_arrived.at(DEPTH))
+        {
+            ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't arrive.");
+            return;
+        }
+    }
+    catch (std::out_of_range)
+    {
+        ROS_DEBUG("Brain: Skipping publish unit vector topic! Depth frame didn't configure.");
+        return;
+    }
 
-		cv::Mat image(height, width, CV_32FC3, data, width * STEP);
+#pragma omp parallel for
+    for (unsigned int y = 0; y < height; ++y) {
+        for (unsigned int x = 0; x < width; ++x) {
+            unit_vectors[y * width * 3 + x * 3] = (static_cast<float>(x) - depth_intrinsics.ppx) / depth_intrinsics.fx;
+            unit_vectors[y * width * 3 + x * 3 + 1] = (static_cast<float>(y) - depth_intrinsics.ppy) / depth_intrinsics.fy;
```

Comment:
In c, the result of the operation between int and float will be float automatically. It is not required to cast an INT before doing the operation. But this cast does not hurt as well since unit vectors are only calculated once per reboot.

Reply:
Yeah, I'm aware of the automatic conversion, I just thought it better C++ code style to cast before doing the arithmetic. It's purely my subjective opinion though, so if others don't like it, I can take it out.